### DISCUSSION TITLE
pending/success/error pattern with jotai -- conditionally check whether `.prop(...)` resolves to undefined or not

### DISCUSSION THREAD

**Discussion by Sebastian-Nielsen** [pmndrs/jotai#2551] [Q&A] [answered]
Im attempting to implement the pending/success/error pattern with jotai. 

I am struggling to create a derived atom for the `deeplyNested` attribute. The error I am getting is understandable: It seems like it is type-checking based on the initialstate in which `data` is `undefined` and so, understandably, the attribute `outer` never exists on the `data` attribute. 

```ts
export type Async<T> = Pending | Success<T> | Error;
type Pending = {
	loadingStatus: 'pending';
	data: undefined;
	error: undefined;
};
type Success<T> = {
	loadingStatus: 'success';
	data: T;
	error: undefined;
};
type Error = {
	loadingStatus: 'error';
	data: undefined;
	error: any;
};

interface Data {
	outer: {
		a: 1,
		b: 2,
		inner: {
			deeplyNested: true
		}
	},
	bar: true
}

export const userDataAtom = atomWithImmer<Async<Data>>({
	loadingStatus: 'pending',
	data: undefined,
	error: undefined,
})

// Goal: create a derived atom for the attribute `deeplyNested`

const x = focusAtom(userDataAtom, optic => {
	return optic.prop('data').prop('outer').prop('inner');
	//                             ^^^^^^^--Error: argument of type 'string' is not assignable to parameter of type 'never'
});
```

Is there a way to conditionally check whether `.prop('data')` resolves to undefined or not?

I would like to return `.prop('outer').prop('inner')` in case it is not undefined, and in case it is undefined I would like to return undefined.

**Comment by rothsandro** [comment, marked as answer]
Just add [`optional()`](https://akheron.github.io/optics-ts/reference-mc/#optional):

```ts
return optic.prop('data').optional().prop('outer').prop('inner');
```
